src/Service/ProductPositionsUpdater.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+declare(strict_types=1);
+
+namespace ThreeBRS\SortingPlugin\Service;
+
+use Doctrine\ORM\EntityManagerInterface;
+use Sylius\Component\Core\Model\ProductTaxonInterface;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Symfony\Component\EventDispatcher\GenericEvent;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Session\Flash\FlashBagInterface;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Contracts\Translation\TranslatorInterface;
+use ThreeBRS\SortingPlugin\Service\Exception\ProductPositionsUpdaterException;
+
+class ProductPositionsUpdater
+{
+    private ?FlashBagInterface $sessionBag = null;
+
+    /**
+     * @var array<int|string, int> Key: PK, Value: index
+     */
+    private array $identifiersPositionsMap = [];
+
+    public function __construct(
+        private EntityManagerInterface $entityManager,
+        private RouterInterface $router,
+        private EventDispatcherInterface $eventDispatcher,
+        private TranslatorInterface $translator,
+    )
+    {
+    }
+
+    /**
+     * @throws ProductPositionsUpdaterException
+     */
+    public function process(Request $request): string
+    {
+        $this->initialize($request);
+
+        if ($this->identifiersPositionsMap === []) {
+            $this->setSessionError('threebrs.ui.sortingPlugin.chooseTaxon');
+        }
+
+        $productTaxonCollection = $this->entityManager
+            ->getRepository(ProductTaxonInterface::class)
+            ->findBy(['id' => array_keys($this->identifiersPositionsMap)]);
+
+        if (count($productTaxonCollection) !== count($this->identifiersPositionsMap)) {
+            throw new ProductPositionsUpdaterException('Required identifiers set is invalid.');
+        }
+
+        $taxon = null;
+
+        foreach ($productTaxonCollection as $productTaxon) {
+            assert($productTaxon instanceof ProductTaxonInterface);
+            $taxon ??= $productTaxon->getTaxon();
+            $productTaxon->setPosition($this->identifiersPositionsMap[$productTaxon->getId()] ?? 0);
+        }
+
+        if ($taxon === null) {
+            $this->setSessionError('threebrs.ui.sortingPlugin.noProductMessage');
+
+            return $this->router->generate('threebrs_admin_sorting_index');
+        }
+
+        $this->entityManager->flush();
+        $this->setSessionSuccess('threebrs.ui.sortingPlugin.successMessage');
+        $this->eventDispatcher->dispatch(new GenericEvent($taxon), 'threebrs-sorting-products-after-persist');
+
+        return $this->router->generate('threebrs_admin_sorting_products', ['taxonId' => $taxon->getId()]);
+    }
+
+    /**
+     * @throws ProductPositionsUpdaterException
+     */
+    private function prepareSessionMessage(string $key): string
+    {
+        if ($this->sessionBag === null) {
+            throw new ProductPositionsUpdaterException('Session\'s state is corrupted.');
```

Comment:
Uzavřel bych to do try catch?

Reply:
Přistoupil jsem k tomu ještě jinak (autorem je Dima, některé řádky vidím poprvé):

- `prepareSessionMessage` vůbec `sessionBag` nepotřebuje a tudíž he nemá ani kontrolovat, není to jeho starost
- obalil jsem do try catch `$this->translator->trans($key)` a případný error zaloguju, protože by bylo nešťastné, kdyby kvůli chybě v překladu padla na fatal hlavní činnost - řazení produktů
- zrušil jsem všechny class properties, které se nastavovaly až metodou `process` - bylo zbytečné koledovat si o zapomenutý stav class properties a divit se při dalším zavolání `process` proč se to chová jinak. Properties jsem převedl na lokální proměnné metody.

Reply:
`$this->translator->trans($key)` nevyhazuje exception, pokud nenajde překlad, tak zobrazí klíč překladu to je vše. Try cacht je tu zbytečně a nic nehlídá.